pkg/machine/fcos.go

```diff
@@ -107,37 +107,50 @@ type FcosDownload struct {
 	Download
 }
 
-func NewFcosDownloader(vmType VMType, vmName string, imageStream FCOSStream, vp VirtProvider) (DistributionDownload, error) {
-	info, err := GetFCOSDownload(vp, imageStream)
+func NewDownload(vmName string, vp VirtProvider) (Download, error) {
+	cacheDir, err := GetCacheDir(vp.VMType())
 	if err != nil {
-		return nil, err
+		return Download{}, err
 	}
-	urlSplit := strings.Split(info.Location, "/")
-	imageName := urlSplit[len(urlSplit)-1]
-	url, err := url2.Parse(info.Location)
+
+	dataDir, err := GetDataDir(vp.VMType())
 	if err != nil {
-		return nil, err
+		return Download{}, err
 	}
 
-	cacheDir, err := GetCacheDir(vmType)
+	return Download{
+		Artifact:        vp.Artifact(),
+		CacheDir:        cacheDir,
+		CompressionType: vp.Compression(),
+		DataDir:         dataDir,
+		Format:          vp.Format(),
+		VMKind:          vp.VMType(),
+		VMName:          vmName,
+	}, nil
+}
+
+func NewFcosDownloader(dl Download, imageStream FCOSStream) (DistributionDownload, error) {
```

Comment:
This should be turned into a method of `Download`

```golang
func (dl *Download) NewFcosDownloader(...)
```

Reasoning is that it's changing the passed `dl` and a method is more expressive than a somehow hidden in/out parameter.